### PR TITLE
faster resbuf

### PR DESCRIPTION
**Pull request by rwcarlsen**
This makes recipe based FCO sims go from ~15 min to ~2 min.  Also I threw in a one line fix for sim_init where prototypes added post simulation init weren't snapshotted properly (the new deployinst utilizes does this and won't work correctly without this fix).

### CODE REVIEW

File: src/context.cc [hunk 1]
```diff
@@ -119,6 +119,7 @@ boost::uuids::uuid Context::sim_id() {
 
 void Context::AddPrototype(std::string name, Agent* p) {
   protos_[name] = p;
+  SimInit::SnapAgent(p);
```
Comment (gidden) [inline src/context.cc]: why is this needed here now?

File: src/toolkit/res_buf.h [hunk 2]
```diff
@@ -306,12 +311,12 @@ class ResBuf {
...
-    typename std::list<typename T::Ptr>::iterator it;
-    std::vector<double> qtys;
-    for (it = rs_.begin(); it != rs_.end(); ++it) {
-      qtys.push_back((*it)->quantity());
+    int n = rs_.size();
+    if (n == 0) {
+      qty_ = 0;
+    } else if (n == 1) {
+      qty_ = rs_.front()->quantity();
     }
```
Comment (gidden) [inline src/toolkit/res_buf.h]: just to confirm, nothing should be done if `rs_.size() > 1`?

Comment (rwcarlsen) [inline src/toolkit/res_buf.h]: correct.

**Comment by rwcarlsen** [comment]
This should be ready.

**Comment by gidden** [comment]
thanks @rwcarlsen!
